src/opt/horizontal_fusion_impl.cpp

```diff
@@ -0,0 +1,839 @@
+            op2_cnt[str] = 1;
+        else
+        {
+            op2_cnt[str] += 1;
+            hash_child = true;
+            break;
+        }
+    }
+    if(hash_child)
+    {
+        // Create a value for this instruction.
+        hash_value& value = create_value(ins);
+        value.set_root();
+        // Flag children to be hashed.
+        for(auto&& output : ins->outputs())
+        {
+            if(op2_cnt[output->name()] > 1)
+                instr2_hash[output] = true;
+        }
+    }
+}
+
+// Find the first axis that matches given dim.
+int horizontal_fusion_impl::find_axis(instruction_ref ins, int dim)
+{
+    int ndx = 0;
+    for(auto&& size : ins->get_shape().lens())
+    {
+        if(size == dim)
+            return ndx;
+        ndx++;
+    }
+    return -1;
+}
+
+// Find axis for convolution filter's output and broadcast's input.
+int horizontal_fusion_impl::find_axis(instruction_ref ins, instruction_ref base, int base_axis)
+{
+    if(is_conv(base))
+    {
+        return get_conv_output_axis();
+    }
+    else if(ins->outputs().at(0)->name() == "broadcast")
+    {
+        if(!match_dim(base, ins->outputs().at(0), -1))
+            MIGRAPHX_THROW("Unmatched output");
+        int dim = base->get_shape().lens().at(base_axis);
+        return find_axis(ins, dim);
+    }
+    else
+        return base_axis;
+}
+
+// Check whether ins1 and ins2 match in all dimensions excluding axis.
+bool horizontal_fusion_impl::match_dim(instruction_ref ins1, instruction_ref ins2, int axis)
+{
+    auto lens1 = ins1->get_shape().lens();
+    auto lens2 = ins2->get_shape().lens();
+    if(lens1.size() != lens2.size())
+        return false;
+    int ndx = 0;
+    for(auto&& size : lens1)
+    {
+        if((size != lens2.at(ndx)) && (ndx != axis))
+            return false;
+        ndx++;
+    }
+    return true;
+}
+
+bool horizontal_fusion_impl::compare_inputs(std::vector<instruction_ref>& input1,
+                                            std::vector<instruction_ref>& input2,
+                                            instruction_ref base_ins,
+                                            int base_axis)
+{
+    if(input1.size() != input2.size())
+        return false;
+    int ndx              = 0;
+    std::size_t base_dim = base_ins->get_shape().lens().at(base_axis);
+
+    for(auto&& ins1 : input1)
+    {
+        instruction_ref ins2 = input2.at(ndx++);
+        if(ins1->name() != ins2->name())
+            return false;
+
+        int axis = find_axis(ins2, base_ins, base_axis);
+        if(axis == -1)
+            return false;
+        if(!match_dim(ins1, ins2, axis))
+            return false;
+        if(ins2->get_shape().lens().at(axis) != base_dim)
+            return false;
+    }
+
+    return true;
+}
+
+void horizontal_fusion_impl::concat(std::vector<instruction_ref>& instrs,
+                                    std::unordered_map<instruction_ref, instruction_ref>& root,
```

Comment:
Both of these parameters can be passed as const.

Reply:
fixed.

Reply:
fixed.